Report (#107)
Itu-DevOps-2026/ITU-MiniTwit: report/report.md

This allows focus on other aspects of the project rather than database administration,
in exchange for some loss of control.

![MiniTwit Deployment Diagram](report/images/MiniTwit_deployment_diagram.png)

## Dependencies

_Authors: Marie, Frederik, Nikolej, Sara_

Besides the .NET packages used to make the MiniTwit application run,
which can be seen [here](#net-dependencies), the most important dependencies can be seen below, categorized by type.

*Note: Some, like Docker, can fall into more than one category.*

![Dependencies](report/images/dependencies.png)

## System States

_Authors: Vitus, Sara_

The project is currently in a strong state with respect to software quality.
From the outset, we established a workflow in which all incoming changes were required to pass automated tests and static analysis checks before being merged into the production branch.
Several analysis tools were also configured with strict settings to enforce high coding standards.
At present, the entire project complies with these quality requirements and can thus be considered stable.

There have been a few necessary compromises to these quality gating rules.

> The project is developed on top of the Chirp! platform, which was originally designed for .NET 8. Since then, newer .NET versions have been released, and .NET 8 is not considered outdated.

Reviewer: ```suggestion
The project is developed on top of the Chirp! platform, which was originally designed for .NET 8. Since then, newer .NET versions have been released, and .NET 8 is now considered outdated.
```

Identifying bottlenecks or failures that would otherwise have been difficult to detect, became significantly easier and made us focus more on maintaining the software in production, rather than only focusing on implementing functionality. Reliability and stability became important parts of the development process rather than something considered only at the end.

**Infrastructure as Code (IaC)**
Luckily, we were never forced to take the whole system down and bring it back online.
Towards the end of the project, we did attempt to ensure `vagrant up` could do exactly that, recreating our infrastructure from scratch.
Ultimately, there were some issues that we did not have time to fix, hence the [pull request #90](https://github.com/Itu-DevOps-2026/ITU-MiniTwit/pull/90) remains open.
A video demo of `vagrant up` can be found [here](report/images/group_e_vagrant_up_demo.gif).

# Use of Generative AI
_Authors: Nikolej, Sara_

During the development of this project we used ChatGPT in three main ways:

**Debugging.** When encountering unexpected bugs and unclear error messages,
we often consulted ChatGPT about possible causes and fixes.

> While it rarely solved issues entirely on its own, it frequently helped narrow down the problem space and suggested relevant debugging strategies.

Reviewer: ```suggestion
While it rarely solved issues entirely on its own, it frequently helped narrow down the problem space.
```

**Research.** This course presents challenges involving numerous technologies,
many of which we were unfamiliar with, including monitoring tools, docker and load balancing.
ChatGPT was used to summarize lengthy documentation, explain unfamiliar concepts, and provide concrete guides tailored to our situation.

**Generating boilerplate / configurations.** ChatGPT was also used for simple, repetitive tasks
such as generating boilerplate code or writing Github Actions Workflow files.

> For example, the `build-report.yml` workflow, that converts the markdown source into a pdf.

Reviewer: ```suggestion
```

After discovering that the application was unavailable, the first step was to inspect the logs. The logs revealed that the outage was caused by thread pool starvation. Further investigation showed that certain requests to the application were taking upwards of 46 minutes. By analyzing the database activity, it became evident that queries were scanning more than 200,000 rows per second. It was therefore decided that the database required proper indexing to mitigate the issue. After the necessary indexes were added and the application was restarted, the system returned to stable operation and performed as expected. The full bug report can be found on issue [#99](https://github.com/Itu-DevOps-2026/ITU-MiniTwit/issues/99)

![Database load before indexing.](report/images/database_fetch.png)

> ## Reflect and describe what was the "DevOps" style of work

Reviewer: ```suggestion
## Reflection and description of "DevOps" style of work
```

# Use of Generative AI
_Authors: Nikolej, Sara_

During the development of this project we used ChatGPT in three main ways:

**Debugging.** When encountering unexpected bugs and unclear error messages,
we often consulted ChatGPT about possible causes and fixes.
While it rarely solved issues entirely on its own, it frequently helped narrow down the problem space.

**Research.** This course presents challenges involving numerous technologies,
many of which we were unfamiliar with, including monitoring tools, docker and load balancing.
ChatGPT was used to summarize lengthy documentation, explain unfamiliar concepts, and provide concrete guides tailored to our situation.

**Generating boilerplate / configurations.** ChatGPT was also used for simple, repetitive tasks
such as generating boilerplate code or writing Github Actions Workflow files.

The use of AI has made these tasks easier, spared us many frustrations during troubleshooting and saved considerable time during development.
At the same time, we met certain limitations when using AI. Suggested fixes sometimes appeared plausible while being incorrect or incompatible with our setup. 

> Additionally, relying on AI summaries may have reduced some of the deeper understanding that can come from manually reading documentation or solving problems independently, like spending hours solving a tiny bug. 

Reviewer: ```suggestion
Additionally, relying on AI summaries may have reduced some of the deeper understanding that can come from manually reading documentation or solving problems independently.
```

Several analysis tools were also configured with strict settings to enforce high coding standards.
At present, the entire project complies with these quality requirements and can thus be considered stable.

There have been a few necessary compromises to these quality gating rules.
The project is developed on top of the Chirp! platform, which was originally designed for .NET 8. Since then, newer .NET versions have been released, and .NET 8 is now considered outdated.
As a result, the Docker Scout analysis consistently recommends upgrading the runtime environment.
Given the project constraints and dependencies on the underlying platform, we chose to explicitly exclude this specific category from the analysis checks. While this represents a deviation from the otherwise strict quality requirements, it is not considered a critical issue as long as the limitation is acknowledged and taken into account during maintenance and future development.

Below is the latest static analysis checks.

![System checks on the recent-most change](report/images/Analysis_checks.png)

> 

Reviewer: ```suggestion
\clearpage
```

**Static Analysis**

- Hadolint - Linter for Dockerfiles
- CSharpier - C# code formatter
- CodeQL - Static analysis engine used to identify security vulnerabilities and code quality issues
- Roslyn - .NET compiler platform for code analysis
- Docker Scout - Security and vulnerability analysis tool for Docker
- SonarCloud - Cloud-based code quality and security analysis platform
- Codacy - Automated code review and quality monitoring platform

**Development**

- Git - Distributed version control system
- GitHub - Web-based platform for hosting Git repositories
- .NET - Microsoft’s development platform and runtime used to make the minitwit application

## System States

_Authors: Vitus, Sara_

The project is currently in a strong state with respect to software quality.
From the outset, we established a workflow in which all incoming changes were required to pass automated tests and static analysis checks before being merged into the production branch.
Several analysis tools were also configured with strict settings to enforce high coding standards.
At present, the entire project complies with these quality requirements and can thus be considered stable.

> Any crashes or critical issues encountered during development were thoroughly investigated, and the underlying causes were addressed accordingly.

Reviewer 2: Delete? This is infered from Maintenance

Reviewer: Resolved, as the comment only applies to line 131. Will manually check.

The setup is pull-based as the application exposes metrics which are then pulled by Prometheus.

In Grafana, the monitoring has been split up into two dashboards; application metrics and infrastructure metrics (see video [Monitoring Dashboards](../README.md:80)).
Application metrics focuses mainly on request rates. It displays CPU usage in seconds, the amount of HTTP request received, and statistics on different types of requests.
Infrastructure metrics focus on the server side and display dashboards containing information about: memory usage, CPU usage and process uptime.

The application's monitoring is at the reactive level [@turnbull2014art],
as only a limited set of primarily operational dashboards are provided, with the main focus being on measuring availability.
The monitoring does not supply data to measure user experience or data to benefit the business side.

There are many ways monitoring could be improved.
For one, database monitoring would have been especially beneficial both for the operational side and to provide metrics for the business side e.g. number of users in the system.

> Lastly, the monitoring dashboards provided by DigitalOcean to monitor the Droplets has been regularly used.

Reviewer 2: ```suggestion

```

Alloy collects logs from running Docker containers through the Docker socket `host = unix:///var/run/docker.sock`,
and populates them with metadata like labels before forwarding to Loki for storage.
Lastly, the logs are queried by Grafana and presented visually.

Our chosen focus on aggregation and visualization is accomplished by grouping logs by containers in Grafana.
The default Grafana Logs Drilldown page is more than sufficient for this purpose,
hence there are no custom dashboards.

By providing a centralized view of all system logs,
this setup improves observability significantly, allowing easier searching and faster response in case of errors.

Even though the logging setup was used in a limited capacity during development,
it proved to be especially useful if for instance the simulation suddenly reports failed requests.
The MiniTwit container logs include errors, HTTP requests and database queries,
which allows us to pinpoint the problem quickly.
Or if the monitoring setup indicates something unexpected,
the aggregated logging platform allows us to ascertain whether it is a problem with the monitoring stack or the application itself.

> 

Reviewer: kan ikke helt regne ud hvad der bliver foreslået skal slettes her. Blank space?

﻿---
title: "ITU-MiniTwit - Group e - Report"
subtitle: |
  DevOps, Software Evolution and Software Maintenance, BSc (Spring 2026)
  **Course code:** BSDSESM1KU
author: |

>   Charlotte Planteig <cpla@itu.dk>  

Reviewer: ```suggestion 
```

  Frederik Hørup Petersen <frap@itu.dk>  
  Marie Johansen <majoh@itu.dk>  
  Nikolej Lundquist <nivl@itu.dk>  

>   Sara Bagger <salb@itu.dk>

Reviewer: ```suggestion
  Marie Højsted Johansen <majoh@itu.dk>  
  Nikolej Vinther Lundquist <nivl@itu.dk>  
  Sara Louise Høg Bagger <salb@itu.dk>
```

![Illustration of code base](docs/images/onion_architecture.png)
_Illustration of the Chirp! app codebase structure - based on onion architecture (reused from the Chirp! project - not compiled during this course)._

## Deployment

_Authors: Nikolej_

Below is a diagram of the overall deployment architecture of the MiniTwit application.
The system uses two load balancers and two production server instances each running all containerized services and
both read and write to the same DigitalOcean Managed Database.

This setup is designed to prioritize availability.
One can imagine the stakeholders of a social media platform pushing for availability,
since every second of downtime is potential earnings lost.

The dual load balancer (LB) setup was chosen for this reason.
The primary load balancer initially handles all traffic, but is replaced by the secondary in case of failure.
This is enabled through the heartbeat messages exchanged between the two using VRRP.
The secondary then takes ownership of the reserved IP and resumes operation.
Essentially, one LB is active while the other is on standby,

> which keeps the system online and reduces one potential single point of failure.

Reviewer: ```suggestion
The dual load balancer (LB) setup was chosen for this reason.
The primary load balancer initially handles all traffic, but is replaced by the secondary in case of failure.
This is enabled through the heartbeat messages exchanged between the two using VRRP.
The secondary then takes ownership of the reserved IP and resumes operation.
Essentially keeping the system online and reduces one potential single point of failure.
```

The same principle is applied to the application layer, by using two MiniTwit production servers.
Both run the exact same services and both are connected to the same database,
which enables traffic to be served by either instance.
This provides redundancy in case one server fails,

> allowing the team to bring it back up while the other instance continues to handle traffic.

Reviewer 2: This as well as line 68 could be moved to Availability and Scalability (maybe). Insert refeence to this section at the end of line 46.
If we could shorten it as well that would be a great start

> However, there is one major drawback to this setup. The monitoring stack is duplicated as well.

Reviewer: ```suggestion
However one major drawback to this setup is the duplicated monitoring stack.
```

Logs and metrics for each server are separate, which compromises observability consistency.
If one server goes down, logs and metrics may be lost.
This could be fixed by having a centralized monitoring stack e.g. on a separate device that all application nodes
export to.

Conversely, since both production serves run identical containerized environments,
the system is easily reproducible and scalable.

While the monitoring and logging data is not strictly centralized, the app database itself is.
DigitalOcean provides a service for database hosting reducing the maintenance burden on the group.
This allows focus on other aspects of the project rather than database administration,
in exchange for some loss of control.

![MiniTwit Deployment Diagram](report/images/MiniTwit_deployment_diagram.png)
_MiniTwit Deployment Diagram_

## Dependencies

_Authors: Marie, Frederik, Nikolej, Sara_

Besides the .NET packages used to make the MiniTwit application run,
which can be seen [here](#net-dependencies).

> The most important dependencies can be seen below, categorized by type.

Reviewer: ```suggestion
which can be seen [here](#net-dependencies), the most important dependencies can be seen below, categorized by type.
```

## Logging

_Authors: Nikolej_

The system uses a minimal push-based logging stack utilizing Alloy [@alloy_docs], Loki [@loki_docs] and Grafana [@grafana_docs],
focusing on aggregation and visualization rather than analysis.
The Grafana-Loki stack was primarily chosen because it is cheap to run
and for its native support for Grafana, which the existing monitoring system was already using.

Alloy collects logs from running Docker containers through the Docker socket `host = unix:///var/run/docker.sock`,
and populates them with metadata like labels before forwarding to Loki for storage.
Lastly, the logs are queried by Grafana and presented visually.

Our chosen focus on aggregation and visualization is accomplished by grouping logs by containers in Grafana.
The default Grafana Logs Drilldown page is more than sufficient for this purpose,
hence there are no custom dashboards.

By providing a centralized view of all system logs,
this setup improves observability significantly, allowing easier searching and faster response in case of errors.

Even though the logging setup was used in a limited capacity during development,

> it proved to be especially useful if for instance the simulation suddenly reports failed requests.

Reviewer 2: ```suggestion
The logging setup proved to be especially useful if for instance the simulation suddenly reports failed requests.
```

The MiniTwit container logs include errors, HTTP requests and database queries,
which allows us to pinpoint the problem quickly.
Or if the monitoring setup indicates something unexpected,

> the aggregated logging platform allows us to ascertain whether it is a problem with the monitoring stack or the application itself.

Reviewer 2: ```suggestion
```

**Static Analysis**

- Hadolint - Linter for Dockerfiles
- CSharpier - C# code formatter
- CodeQL - Static analysis engine used to identify security vulnerabilities and code quality issues
- Roslyn - .NET compiler platform for code analysis
- Docker Scout - Security and vulnerability analysis tool for Docker
- SonarCloud - Cloud-based code quality and security analysis platform
- Codacy - Automated code review and quality monitoring platform

**Development**

- Git - Distributed version control system
- GitHub - Web-based platform for hosting Git repositories
- .NET - Microsoft’s development platform and runtime used to make the minitwit application

## System States

_Authors: Vitus, Sara_

The project is currently in a strong state with respect to software quality.
From the outset, we established a workflow in which all incoming changes were required to pass automated tests and static analysis checks before being merged into the production branch.
Several analysis tools were also configured with strict settings to enforce high coding standards.
At present, the entire project complies with these quality requirements and can thus be considered stable.

> Any crashes or critical issues encountered during development were thoroughly investigated, and the underlying causes were addressed accordingly.

Reviewer: ```suggestion
```
